libethash/internal.c

```diff
@@ -400,38 +401,48 @@ ethash_full_t ethash_full_new_internal(
 	ret->file_size = (size_t)full_size;
 	switch (ethash_io_prepare(dirname, seed_hash, &f, (size_t)full_size, false)) {
 	case ETHASH_IO_FAIL:
+		// ethash_io_prepare will do all ETHASH_CRITICAL() logging in fail case
 		goto fail_free_full;
 	case ETHASH_IO_MEMO_MATCH:
 		if (!ethash_mmap(ret, f)) {
+			ETHASH_CRITICAL("mmap failure()");
 			goto fail_close_file;
```

Comment:
why? goto is a very common error checking pattern in C code, in order to maintain a single function exit strategy and RAII. Check [Chapter 7 of the Kernel coding style](https://www.kernel.org/doc/Documentation/CodingStyle)